Performance: Faster document title & content search with full-text indexes

Merinorus opened the pull request:
## Proposed change

This PR basically adds a full-text index for Title & Content searches on all supported databases: PostgreSQL, MariaDB, and SQLite.  The goal is to have a "title & content search" that scales better with a high number of documents. The user can still choose between `Title & Content` or `Advanced Search`.

Full-text indexes are added for the `title` and `content` columns of the document table, as they may contain thousands or tens of thousands of rows. For tags, correspondents, etc., the current full scan (`icontains`) is simple and probably performant enough, except in some edge cases (hundreds of tags?).

I tried to keep the syntax as close as possible to the Django ORM, but behind the scenes, it behaves differently depending on the database backend.

The global search also benefits from this improvement.

I'm not a Django specialist nor a DB expert. I checked that the full text is correctly used with `EXPLAIN` SQL queries and manual benchmarking. There is no need to hurry for this PR; any feedback is very welcome!


Here are some discussions about search performance issues:
- [High RAM Usage, Slow Response Times On Search #4503](https://github.com/paperless-ngx/paperless-ngx/discussions/4503)
- [Suboptimal SQL queries makes UI hard to use with postgres and > 10 000 documents #4922](https://github.com/paperless-ngx/paperless-ngx/discussions/4922)
- [Very slow queries causing very slow search #5866](https://github.com/paperless-ngx/paperless-ngx/discussions/5866)
- [Search and documents way too slow #9649](https://github.com/paperless-ngx/paperless-ngx/discussions/9649)

## Limitations:
-  In PostgreSQL, due to how the stemming is done, fewer results will be found than with a true `ILIKE %A%` search. Especially partial email address may not be found. Example with  `john@gmail.com`: `john@gmai`, `john@gmail`, and `john@gmai.c` won't work, but `john`, `john@`, `john@gmai.co`, and `john@gmai.com` will.
- This implementation doesn't support CJK characters: in this case, it fallbacks to the non-indexed `icontains` search.


## Type of change


- [x] Bug fix: non-breaking change which fixes an issue.
- [x] New feature / Enhancement: non-breaking change which adds functionality. _Please read the important note above._
- [ ] Breaking change: fix or feature that would cause existing functionality to not work as expected.
- [ ] Documentation only.
- [ ] Other. Please explain:

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [ ] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

Merinorus committed:
Indexed title-content SQL search

Merinorus committed:
Rework RAW SQL fulltext search as a custom Django lookup

Merinorus committed:
Global search: use DB index instead of Whoosh index

shamoon commented:
Thanks for this, and your work in general. Improving search performance is a worthy goal, and we appreciate the effort. That said, I have some pretty deep reservations about this approach, and wanted to flag them before you invest too much more time:

- Feature mismatch: This removes Whoosh from global search, but doesn’t replicate key features like all of the whoosh syntax stuff (`tag:xyz` etc), query correction and more. So in some ways, this is a breaking change that might _worsen_ results.
- Maintainability: Embedding DB-specific SQL inside Django ORM lookups seems kinda brittle and I worry will cause issues across DB versions and future Django upgrades. Not to mention testing across DBs. This in particular makes me extremely hesitant to adopt something like this into our codebase.
- SQLite: The FTS5 virtual table with wide prefix indexes can massively increase disk usage — that needs more analysis. Not to mention the impact of all of this on backup/restore. I sometimes dream of dropping SQLite support, in some ways it would free us up, but that is of course not a choice to be made lightly (or here).
- No benchmarking / performance / idea of impact: It would be helpful to have data showing performance wins across different DBs and dataset sizes. Massive installs are frankly not the most common situation for paperless users, and moreover Im not convinced they should be our target audience. So what is the 'cost' to something like this for most users, and for what gain (any?)? And again, how does it perform across db types, system constraints, etc? SQLite with FTS5 on an RPi?

So again, I’d be hesitant to move forward with this as-is. Perhaps as an opt-in kinda thing, but I’m not convinced this should replace Whoosh (or anything else) without a more complete plan/idea of its impact. Frankly, as we have long-discussed, especially as it seems to have lost much of its OSS momentum, the idea of entirely [replacing whoosh with some other FTS backend](https://github.com/paperless-ngx/paperless-ngx/discussions/7648) might be more appealing.

Of course just my opinion, and hope it goes without saying Im open to discussing more. Thanks again for all your contributions.

Merinorus committed:
Fulltext search fixes and improvements

- Fix wrong SQL requests
- Revert to non-indexed search for CJK characters (full-text not supported)
- SQLite migration: contentless virtual table and no prefix (about +10% total file size)
- Clearer comments

Merinorus committed:
Global search: reintroduce setting to search on Whoosh or FT index

Merinorus commented:
Thank you for your kind words, I appreciate..
I'll try to answer point by point:

- Feature mismatch: I didn't know the Whoosh syntax could be used in the global search, my bad, I reenabled it on the flag (`db_only = False`). Plus, it might be the only solution currently to search in some languages such as Chinese and Japanese (see below).
- Maintainability: You’re totally right. Custom SQL queries require careful testing. Plus, the current CI doesn’t offer integration tests across various databases, so regressions might happen during a database version upgrade. We could implement that (it’s a lot of effort), or limit the number of custom queries. At first, I wrote raw SQL directly in the filter, but it’s not maintainable. Personally, I find that implementing Django custom functions or lookups in a dedicated file would be clear enough. Once tested, hopefully upgrading the database rarely breaks behavior, and fixing it would be manageable.
- SQLite: Indeed, it was about 70% bigger! I checked at the SQLite documentation to make a contentless FTS table without prefix, now it's only 10% bigger, and the full-text search still works fast enough.
- Performance / impact: In my setup (2700 documents), the full-text index makes the search 5 to 10 times faster than without index (from 5-15s to 1-1.5s). The more the document and the number of characters, the more **exponential** the difference becomes.

When you speak about SQLite, I wonder if we shouldn't stick to only one database backend: Django always had a better support of PostgreSQL, and it would simplify working on the DB. I've no personal preference between MariaDB and PostgreSQL, but specifically in Django projects, for advanced text search features and for long term support, I would chose PostgreSQL.


I have no experience with Meilisearch or other alternatives. It might be appropriate, but who knows whether it won’t be abandoned like Whoosh? Maintaining database queries seems easier to me (a bit more code, but less infrastructure, and SQL won’t disappear anytime soon). I can help, but your opinion on this matters more: you are one of the active maintainers doing most of the work, so it’s reasonable to stick with a solution that you can maintain easily. I’ve read the [discussion]([https://github.com/paperless-ngx/paperless-ngx/discussions/7648]) you mentioned: **Tantivy seems promising**.


For performance impact, here is the result on my setup:
- Intel N3710 (a weak 4-core Atom CPU from 2016), 8 GB RAM, 5400 rpm HDD with SSD read cache
- PostgreSQL 17, hosting 2700 documents, and 30 million characters.

Here are the results for a few searches. I just put the length of the alphanumerical keywords, as they are personal. Note that this is a rough estimate, as my CPU takes some time just to render Django views:

| keyword             | Before | After | Advanced search |
| ------------------- | ------ | ----- | --------------- |
| xxxxx               | 6.5 s  | 1.3 s | 2.5 s           |
| xxxx                | 5.5 s  | 1.2 s | 1.1 s           |
| xxxxxx              | 12.0 s | 1.1 s | 4.0 s           |
| xxxxxxxxx           | 11.0 s | 1.6 s | 2.8 s           |
| xxxxxxxx xx xxxxxxx | 13.5 s | 1.9 s | 1.6 s           |
| xx xxxxx            | 8.0 s  | 2.0 s | 3.5 s           |

The advanced search performance is still “okay” for me (I can find it in less than 10 seconds), and it’s convenient because of the various filters. For a simple title or content search, I’d say that properly indexed SQL queries will roughly outperform Whoosh starting from a few thousand documents.  
Of course, I’m speaking only about title and content search — Whoosh brings very useful search features.

To check the other implementations, I exported my Paperless-ngx instance to MariaDB and SQLite databases and tested them as well.

I evaluated all queries by measuring execution time and reviewing `EXPLAIN` clauses. It’s very long, but I can provide them if needed.

All the implementations seem to work, but their behavior is different from `LIKE %A%`. Some results are dropped.  
The other big issue is… CJK characters.  
Since in many Asian languages, words are not separated by spaces, the databases don’t support proper full-text search.

I still have to test Groonga, which seems to integrate easily with PostgreSQL and MariaDB (not MySQL, unfortunately).

For SQLite, Signal made a [custom tokenizer](https://github.com/signalapp/Signal-FTS5-Extension) , but it doesn't seem easy to integrate it with Django.

**To conclude**: Indeed, this PR has some issues! It is hard to support multiple database backends, and focusing on supporting only one (eg: PostgreSQL) would help a lot for maintainability. **Focusing on improving advanced search may be a more viable alternative**.

I'll keep it as a draft for the moment, if someone want to tinker with it feel free to do it. If you think it has to be closed, that's totally fine. I learnt a lot on this one, it was worth exploring.
